src/main/java/org/cyclonedx/gradle/CycloneDxTask.java

```diff
@@ -362,85 +373,106 @@ public void createBom() {
+                new org.cyclonedx.model.Dependency(generatePackageUrl(pluginProject));
+        dependencies.put(generatePackageUrl(pluginProject), rootDependency);
+        Stream.concat(Stream.of(pluginProject), pluginProject.getSubprojects().stream())
+                .forEach(project -> processProject(project, rootDependency, dependencies, builtDependencies));
+        writeBom(metadata, new HashSet<>(components.values()), dependencies.values());
+    }
 
-        Set<Project> projectsToScan = Stream.concat(Stream.of(rootProject), rootProject.getSubprojects().stream())
-                .filter(p -> !shouldSkipProject(p))
-                .collect(Collectors.toSet());
+    private void processProject(
+            Project project,
+            org.cyclonedx.model.Dependency rootDependency,
+            Map<String, org.cyclonedx.model.Dependency> dependencies,
+            Set<String> builtDependencies) {
+        getLogger().debug("Processing project [{}]", project.getName());
+        if (shouldSkipProject(project)) {
+            getLogger().debug("Project [{}] skipped due to plugin configuration", project.getName());
+            return;
+        }
+        String projectReference = generatePackageUrl(project);
+        org.cyclonedx.model.Dependency projectDependency = new org.cyclonedx.model.Dependency(projectReference);
+        for (Configuration configuration : project.getConfigurations()) {
+            processConfiguration(configuration, projectReference, projectDependency, dependencies, builtDependencies);
+        }
+
+        if (!getProject().equals(project)) {
+            rootDependency.addDependency(projectDependency);
+            // declare sub-project as component
+            Component component = generateProjectComponent(project);
+            String bomRef = component.getBomRef();
+            components.putIfAbsent(bomRef, component);
+        }
+    }
 
-        projectsToScan.forEach(project -> {
-            Set<Configuration> configurations = project.getConfigurations().stream()
-                    .filter(configuration -> shouldIncludeConfiguration(configuration)
-                            && !shouldSkipConfiguration(configuration)
-                            && DependencyUtils.canBeResolved(configuration))
+    private void processConfiguration(
+            Configuration configuration,
+            String projectReference,
+            org.cyclonedx.model.Dependency moduleDependency,
+            Map<String, org.cyclonedx.model.Dependency> dependencies,
+            Set<String> builtDependencies) {
+        getLogger().debug("Processing configuration [{}]", configuration.getName());
+        if (!shouldIncludeConfiguration(configuration) || shouldSkipConfiguration(configuration)) {
+            getLogger().debug("Configuration [{}] skipped due to plugin configuration", configuration.getName());
+            return;
+        }
+        if (!configuration.isCanBeResolved()) {
```

Comment:
Is there a difference between `DependencyUtils.canBeResolved(configuration))` and `configuration.isCanBeResolved()`?

Reply:
`DependencyUtils.canBeResolved(configuration))` is using reflection to maintain compatibility with gradle v<3  https://github.com/CycloneDX/cyclonedx-gradle-plugin/blob/master/src/main/java/org/cyclonedx/gradle/utils/DependencyUtils.java#L40